src/engine/renderer/tr_shader.cpp

```diff
@@ -2895,26 +2969,83 @@ static bool ParseStage( shaderStage_t *stage, const char **text )
-		// normalScale <float> <float> 
-		//
-		// scale normalMap channel, can flip channels with negative values
-		//
-		// expects OpenGL scale as default: 1 1 1
-		//
-		// OpenGL format is described there:
-		//     https://github.com/KhronosGroup/glTF/tree/2.0/specification/2.0#materialnormaltexture
-		//
-		// > The normal vector uses OpenGL conventions where +X is right, +Y is up, and +Z points toward the viewer.
-		//
-		// examples of scales:
-		//
-		//  1   1   1    OpenGL format (default)
-		//  1  -1   1    DirectX format with reverted green channel (reverted Y component)
-		// -1   1   1    weird other format with reverted red channel (reverted X component)
-		//  1   1   1.5  OpenGL format with an increased blue channel (larger Z componnent)
+		/* normalFormat <[-]X> <[-]Y> <[-]Z>
+
+		Describes the normal map format for the the given normal map file.
+
+		Since this parser assumes DirectX format with normalMap keyword
+		to keep compatibility with materials created for ioquake3
+		normalFormat must be set to: X Y Z
+		to load normal maps in OpenGL format using normalMap keyword.
+
+		OpenGL format is described there:
+		  https://github.com/KhronosGroup/glTF/tree/2.0/specification/2.0#materialnormaltexture
+
+		> The normal vector uses OpenGL conventions where +X is right, +Y is up, and +Z points toward the viewer.
+
+		examples of formats:
+
+		X  Y  Z OpenGL format
+		X -Y  Z DirectX format with reverted green channel (reverted Y component)
+		-X  Y  Z weird other format with reverted red channel (reverted X component)
+
+		Unlike normalScale that can be used to revert channels by setting them
+		negative, the normalFormat keyword is meant to be engine agnostic:
+		engines implementing either OpenGL or DirectX format internally
+		are expected to do the required normal channel flip themselves.
+		*/
+		else if ( !Q_stricmp( token, "normalFormat" ) )
+		{
+			const char* components[3] = { "X", "Y", "Z" };
+			vec3_t normalFormat;
+
+			for ( int i = 0; i < 3; i++ )
+			{
+				token = COM_ParseExt( text, false );
+
+				if ( !Q_stricmp( token, components[ i ] ) )
+				{
+					normalFormat[ i ] = 1.0;
+				}
+				else if ( token[ 0 ] == '-' && !Q_stricmp( token + 1, components[ i ] ) )
+				{
+					normalFormat[ i ] = -1.0;
+				}
+				else
+				{
+					if ( token[ 0 ] == '\0' )
+					{
+						Log::Warn("missing normalFormat parm in shader '%s'", shader.name );
+					}
+					else
+					{
+						Log::Warn("unknown normalFormat parm in shader '%s': '%s'", shader.name, token );
+					}
+
+					break;
+				}
+
+				if ( i == 2 )
+				{
+					// If all three components are read properly,
+					// set normal format, replace existing one if exists.
+					SetNormalFormat( stage, normalFormat, true );
+				}
+			}
+
+			SkipRestOfLine( text );
+			continue;
+		}
+		/* normalScale <float> <float> <float>
+
+		Scales normal map channels.
+
+		For compatibility purpose, ioquake3 renderer2 syntax is supported:
+		normalScale <float> <float>
```

Comment:
What is the intended effect of this when the z component is reconstructed?

Reply:
when z component is reconstructed, normalformat is not used but normalscaled is used, but here it's 1.0 and it's a factor, it will not revert anything because positive, and do nothing because 1.

Reply:
note: this may be confusing, but that's the comment for the normalscale not normalformat, I edited comments.